sw/device/lib/crypto/impl/rsa/rsa_padding.c

```diff
@@ -106,13 +150,41 @@ static status_t digest_info_write(const otcrypto_hash_digest_t message_digest,
       memcpy(encoding + kSha512DigestWords, &kSha512DigestIdentifier,
              sizeof(kSha512DigestIdentifier));
       break;
+    case kOtcryptoHashModeSha3_224:
+      if (message_digest.len != kSha3_224DigestWords) {
+        return OTCRYPTO_BAD_ARGS;
+      }
+      memcpy(encoding + kSha3_224DigestWords, &kSha3_224DigestIdentifier,
+             sizeof(kSha3_224DigestIdentifier));
+      break;
+    case kOtcryptoHashModeSha3_256:
+      if (message_digest.len != kSha256DigestWords) {
+        return OTCRYPTO_BAD_ARGS;
+      }
+      memcpy(encoding + kSha3_256DigestWords, &kSha3_256DigestIdentifier,
+             sizeof(kSha3_256DigestIdentifier));
+      break;
+    case kOtcryptoHashModeSha3_384:
+      if (message_digest.len != kSha3_384DigestWords) {
+        return OTCRYPTO_BAD_ARGS;
+      }
+      memcpy(encoding + kSha3_384DigestWords, &kSha3_384DigestIdentifier,
+             sizeof(kSha3_384DigestIdentifier));
+      break;
+    case kOtcryptoHashModeSha3_512:
+      if (message_digest.len != kSha3_512DigestWords) {
+        return OTCRYPTO_BAD_ARGS;
+      }
+      memcpy(encoding + kSha3_512DigestWords, &kSha3_512DigestIdentifier,
+             sizeof(kSha3_512DigestIdentifier));
+      break;
     default:
       // Unsupported or unrecognized hash function.
       return OTCRYPTO_BAD_ARGS;
   };
 
   // Copy the digest into the encoding, reversing the order of bytes.
-  for (size_t i = 0; i < message_digest.len / 2; i++) {
+  for (size_t i = 0; i < ceil_div(message_digest.len, 2); i++) {
```

Comment:
The previous version of this loop termination condition contained a subtle assumption that the length of the hash digest is an even number of 32-bit words, which is not true for SHA3-224, which has 28-byte digests (7 words). The `ceil_div` call ensures we do not forget to write the middle word of a hash digest with an odd number of words.

Actually, I'm not entirely convinced the `tmp` variable inside the loop is strictly necessary. @jadephilipoom can you chime in on this?

Reply:
Good catch on the assumption; I think this was adapted from a byte-reversal-in-place implementation. As for `tmp`, it's not strictly necessary since this isn't in-place, so fine with me to factor it out.